### PR TITLE
8267893: Improve jtreg test failure handler do get native/mixed stack traces for cores and live processes

### PR DESCRIPTION
**Pull request by lmesnik**
EFH is improved to process cores and get mixed stack traces with jhsdb and native stack traces with gdb/lldb. It might be useful because hs_err doesn't contain info about all threads, sometimes it is even not generated.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267893](https://bugs.openjdk.java.net/browse/JDK-8267893): Improve jtreg test failure handler do get native/mixed stack traces for cores and live processes

### Reviewers
 * [Igor Ignatyev](https://openjdk.java.net/census#iignatyev) (@iignatev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4234/head:pull/4234` \
`$ git checkout pull/4234`

Update a local copy of the PR: \
`$ git checkout pull/4234` \
`$ git pull https://git.openjdk.java.net/jdk pull/4234/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4234`

View PR using the GUI difftool: \
`$ git pr show -t 4234`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4234.diff">https://git.openjdk.java.net/jdk/pull/4234.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Hi Leonid,

What is EFH? Please update the bug and PR to explain.

Thanks,
David

**Comment by lmesnik** [comment]
> Hi Leonid,
> 
> What is EFH? Please update the bug and PR to explain.
> 
> Thanks,
> David

Updated summary to "Improve jtreg test failure handler do get native/mixed stack traces for cores and live processes".

File: test/failure_handler/src/share/conf/common.properties [hunk 1]
```diff
@@ -35,6 +35,7 @@ onTimeout=\
         jcmd.vm.symboltable jcmd.vm.uptime jcmd.vm.dynlibs \
         jcmd.vm.system_properties \
   jcmd.gc.heap_info jcmd.gc.class_histogram jcmd.gc.finalizer_info \
+  jcmd.vm.info \
```
Comment (iignatev) [inline test/failure_handler/src/share/conf/common.properties:38]: this seems to be unrelated to the RFE you are working on.

File: test/failure_handler/src/share/classes/jdk/test/failurehandler/ToolKit.java [hunk 2]
```diff
@@ -51,6 +56,26 @@ public void gatherEnvironmentInfo(HtmlSection section) {
...
+    public void gatherCoreInfo(HtmlSection section, Path core) {
+        if (core.getFileName().toString().endsWith(".gz")) {
+            Path unpackedCore = Path.of(core.toString().replace(".gz", ""));
+            try (GZIPInputStream gzis = new GZIPInputStream(Files.newInputStream(core))) {
+                Files.copy(gzis, unpackedCore);
+                for (ActionSet set : actions) {
+                    set.gatherCoreInfo(section, unpackedCore);
+                }
+            } catch (IOException ioe) {
+                // just silently skip gzipped core processing
```
Comment (iignatev) [inline test/failure_handler/src/share/classes/jdk/test/failurehandler/ToolKit.java:69]: we don't silently ignore exceptions in `failure_handler`, we tend to print them so we at least have some echoes of what happened.

File: test/failure_handler/src/share/conf/common.properties [hunk 3]
```diff
@@ -60,6 +62,10 @@ jstack.app=jstack
 jstack.args=-e -l %p
 jstack.params.repeat=6
 
+cores=jhsdb.jstack
+jhsdb.jstack.app=jhsdb
+jhsdb.jstack.args=jstack --mixed --core %p --exe %java
```
Comment (iignatev) [inline test/failure_handler/src/share/conf/common.properties:67]: strictly speaking, we can't guarantee that the executable is always `bin/java`, but it's the most common and the most interesting case, so this assumption is good enough for our pourposes.

could you please add a comment explaining this so the further reading won't be puzzled?

File: test/failure_handler/src/share/classes/jdk/test/failurehandler/action/PatternAction.java [hunk 4]
```diff
@@ -59,6 +59,9 @@ public ProcessBuilder prepareProcess(HtmlSection section,
         for (int i = 0, n = args.length; i < n; ++i) {
             args[i] = args[i].replace(pattern, value) ;
         }
+        for (int i = 0, n = args.length; i < n; ++i) {
+            args[i] = args[i].replace("%java", helper.findApp("java").getAbsolutePath());
```
Comment (iignatev) [inline test/failure_handler/src/share/classes/jdk/test/failurehandler/action/PatternAction.java:63]: are we sure that `java` from `PATH` (which is what `findApp` returns) is the right `java`?

File: test/failure_handler/src/share/classes/jdk/test/failurehandler/ToolKit.java [hunk 5]
```diff
@@ -51,6 +56,26 @@ public void gatherEnvironmentInfo(HtmlSection section) {
...
+            }
+            unpackedCore.toFile().delete();
```
Comment (iignatev) [inline test/failure_handler/src/share/classes/jdk/test/failurehandler/ToolKit.java:71]:
```suggestion
            Paths.delete(unpackedCore);
``` ?

File: test/failure_handler/src/share/classes/jdk/test/failurehandler/Utils.java [hunk 6]
```diff
@@ -66,6 +69,16 @@ public static Properties getProperties(String name) {
         String resourceName = String.format(
                 "/%s.%s", name.toLowerCase(), "properties");
         InputStream stream = Utils.class.getResourceAsStream(resourceName);
+
+        // EFH_CONF_DIR might re-defined to load custom configs for development purposes
```
Comment (iignatev) [inline test/failure_handler/src/share/classes/jdk/test/failurehandler/Utils.java:73]: this also seems to be unrelated to the subject and does require documentation in `test/failure_handler/README`.

File: test/failure_handler/Makefile [hunk 7]
```diff
@@ -38,7 +38,7 @@ SOURCES := ${SRC_DIR}/jdk/test/failurehandler/*.java                   \
 
 CONF_DIR = src/share/conf
 
-JAVA_RELEASE = 7
```
Comment (iignatev) [inline test/failure_handler/Makefile:41]: could you please update `DEPENDENCES` section in `test/failure_handler/README`?

File: test/failure_handler/src/share/conf/mac.properties [hunk 8]
```diff
@@ -61,9 +63,21 @@ native.stack.delimiter=\0
 native.stack.params.repeat=6
 native.stack.args=-c\0DevToolsSecurity --status | grep -q enabled && lldb -o 'attach %p' -o 'thread backtrace all' -o 'detach' -o 'quit'
 
+native.jhsdb.app=bash
+native.jhsdb.jstack.delimiter=\0
+native.jhsdb.jstack.args=-c\0DevToolsSecurity --status | grep -q enabled && jhsdb jstack --mixed --pid %p
```
Comment (iignatev) [inline test/failure_handler/src/share/conf/mac.properties:68]: AFAICS `jhsdb` does check "Developer mode" on macos before trying to attach and will just report 'lack of privileges' (as opposed to hanging with a modal window asking for permission), so you can move `jshdb.jstack` to `common.properties`.

**Review by iignatev — CHANGES_REQUESTED** [review]
@lmesnik , how has this been tested?

File: test/failure_handler/Makefile (hunk 7 above)
Comment (lmesnik) [reply test/failure_handler/Makefile:41]: done

**Comment by lmesnik** [comment]
> @lmesnik , how has this been tested?

I used it in the loom for several weeks.

**Comment by lmesnik** [comment]
@iignatev, thank you for your comments. I updated all of them.

File: test/failure_handler/src/share/classes/jdk/test/failurehandler/GathererFactory.java [hunk 9]
```diff
@@ -26,7 +26,10 @@
 import jdk.test.failurehandler.action.ActionHelper;
 import jdk.test.failurehandler.value.InvalidValueException;
 
+import java.io.File;
+import java.io.FileWriter;
 import java.io.PrintWriter;
+import java.nio.file.Files;
```
Comment (iignatev) [inline test/failure_handler/src/share/classes/jdk/test/failurehandler/GathererFactory.java:32]: I don't see why we need these 3 new imports.

File: test/failure_handler/src/share/classes/jdk/test/failurehandler/ToolKit.java [hunk 10]
```diff
@@ -25,20 +25,27 @@
 
 import jdk.test.failurehandler.action.ActionSet;
 import jdk.test.failurehandler.action.ActionHelper;
+import jdk.test.failurehandler.action.PatternAction;
```
Comment (iignatev) [inline test/failure_handler/src/share/classes/jdk/test/failurehandler/ToolKit.java:28]: redundant import

File: test/failure_handler/src/share/conf/linux.properties [hunk 11]
```diff
@@ -55,6 +56,12 @@ native.stack.params.repeat=6
 # has to be the last command
 native.core.app=kill
 native.core.args=-ABRT %p
+
+cores=native.gdb
+native.gdb.app=gdb
+native.gdb.args=%java\0-c\0%p\0-batch\0-ex\0thread apply all backtrace
```
Comment (iignatev) [inline test/failure_handler/src/share/conf/linux.properties:62]: could you please add a comment similar to the one in `common.properties` file?

File: test/failure_handler/src/share/conf/mac.properties [hunk 12]
```diff
@@ -64,6 +64,13 @@ native.stack.args=-c\0DevToolsSecurity --status | grep -q enabled && lldb -o 'at
 # has to be the last command
 native.core.app=kill
 native.core.args=-ABRT %p
+
+cores=native.lldb
+native.lldb.app=lldb
+native.lldb.delimiter=\0
+native.lldb.args=--core\0%p\0%java\0-o\0thread backtrace all\0-o\0quit
```
Comment (iignatev) [inline test/failure_handler/src/share/conf/mac.properties:71]: could you please add a comment similar to the one in common.properties file?

File: test/failure_handler/src/share/conf/mac.properties [hunk 13]
```diff
@@ -64,6 +64,13 @@ native.stack.args=-c\0DevToolsSecurity --status | grep -q enabled && lldb -o 'at
...
+native.lldb.params.timeout=3600000
```
Comment (iignatev) [inline test/failure_handler/src/share/conf/mac.properties:72]: why does `lldb` require an increases timeout, but `gdb` and `jhsdb` do not?

**Review by iignatev — CHANGES_REQUESTED** [review]

**Review by iignatev — APPROVED** [review]
